### ISSUE TITLE
Performance Profiling & Testing

### ISSUE THREAD

**Issue by kwridan** [tuist/tuist#820]
## Need/problem
Introduce a mechanism to aid profiling and performance testing Tuist

## Motivation
We’ve noticed the generation process can take upward of a minute in one of our use cases which we feel is on the slower side and would like to investigate if there are any improvements we can make.

The generation process needs to be fairly fast. Especially when moving to a workflow where the projects are no longer checked into the repository, developers would need to re-generate the projects more frequently.

It’s unrealistic to attempt to say _”it needs to be **n** seconds”_ as of course workspaces come in all sort of shapes and sizes! 

## Detailed design
To aid this effort a few steps can be taken (incrementally):

- [ ] Add a large fixture generator
- [ ] Profile Tuist using a large fixture 
- [ ] Add a custom GitHub Action to perform automated benchmarking and posting results on PRs
- [ ] (Possibly) Once this process is stable a threshold can be locked in to fail builds that exceed it

### Large Fixture

Adding a large fixture manually will be hard to maintain on the long run as any updates needed (e.g. to the manifests or its structure) could get quite tedious. Instead we could write a small fixture generator (separate command line tool) which we can configure. 

For example we can configure it to generate a fixture that has 10 projects, with 5 targets, each with 100 source files or perhaps something more extreme like projects with 1000s of sources.

Initially this fixture generator could live in the same repository as Tuist, it can live within `fixtures/stress_test/FixtureGenerator`  and can move out to its own repo if needed.

### GitHub Actions

We can write a GitHub Action that performs the following tasks

- Generate a pre-defined large fixture
- Pull latest changes from branch
- Build Tuist (with release configuration)
- Measure the time taken to generate the large fixture *(potentially a few runs to average the results)*

Those steps can be repeated once for master and once for the PR branch

The results can then be posted as a comment on the PR.

e.g.

| Fixture  | master | PR 123 | Delta |
| ------------- | ------------- |----|---|
| Large Fixture (1000 sources)  | 30 seconds | 60 seconds | +30 seconds |
| Large Fixture (1000 targets)  | 60 seconds | 120 seconds | +60 seconds |
| Large Fixture (100 projects)  | 120 seconds | 100 seconds | -20 seconds |
| Fixture with swift packages | 120 seconds | 200 seconds | +80 seconds |
| ... |  |  |   |

Some prior art:
[matchai/criterion-compare-action](https://github.com/matchai/criterion-compare-action)

# Drawbacks
- Introducing new GitHub Actions isn’t free, we’d need to continue to maintain this custom action - something we may not all be very familiar with
- PR checks will take even longer to complete with the introduction of another check
	- We could test out the impact by initially making this check optional such that it gets triggered via a special comment (e.g. **“tuist benchmark this”**)

## Alternatives
- Add more performance integration tests at the component level (e.g. [GraphCircularDetectorTests.test_performance](https://github.com/tuist/tuist/blob/master/Tests/TuistCoreTests/Utils/GraphCircularDetectorTests.swift#L170)). 

	- This proposal doesn’t replace / prevent adding more granular performance tests to individual components as needed - the key point is to lock it in or raise awareness of regressions

- Instead of performing benchmarking on a single large fixture, we could run benchmarks on existing fixtures
	- In fact this may offer more variety in addition to the large fixture as some fixture leverage various tuist features (e.g. pre-compiled libraries, swift package dependencies etc…) 
	
- Instead of a custom GitHub Action, we could potentially tweak the current acceptance tests to include some additional logic to collect generation times and post them as a comment on the PR
	- Some work is still needed to perform the measurements on master as times on their own without a reference point aren’t useful
	- Storing master’s results may not yield to accurate comparisons as different CI runs could run on different virtual images with different capabilities

# How we teach this
The development guidelines would need to be updated to document how any of the new tools / actions / scripts work together to aid others in enhancing and maintaining it.

Additionally, as we profile Tuist and make improvements we can introduce a performance guidelines document where we can capture some of the lessons learnt. This could be helpful to ensure we keep those points in mind while developing new features (e.g. avoiding/minimising disk IO / shell calls etc…)

# Unresolved questions
- I’m personally not as familiar with GitHub Actions so will be learning as I go, of course contributions / assistance with this would be much appreciated and welcome.
- Are there any existing benchmarking tools we could leverage?

# References

Related Issues:
- https://github.com/tuist/tuist/issues/329

Somewhat related but looks like its mostly geared towards benchmarking Swift itself:
- [swift/benchmark](https://github.com/apple/swift/tree/master/benchmark)
	- [Contributing to the Swift Benchmark Suite · Patrick Balestra](https://patrickbalestra.com/blog/2018/11/12/contributing-to-the-swift-benchmark-suite.html)

Custom GitHub Actions for benchmarking:
- [matchai/criterion-compare-action](https://github.com/matchai/criterion-compare-action)

**Comment by pepibumur** [comment]
Hey @kwridan. Thanks for bringing this up. This is another matter that I believe is the right time to tackle. I think @andreacipriani is also interested in contributing to this one to speed up the generation of projects at SoundCloud.

> To aid this effort a few steps can be taken (incrementally):

I agree with the execution that you are suggesting here. Having a tripwire that fails PRs when they make the generation slower is crucial to prevent regressions.

> The results can then be posted as a comment on the PR.

Wow! This is neat

> Introducing new GitHub Actions isn’t free, we’d need to continue to maintain this custom action - something we may not all be very familiar with

I think the cost is worth it. We have to maintain other GitHub actions anyways so the knowledge already exists in the organization and project.

> PR checks will take even longer to complete with the introduction of another check 

I think it's ok if we are talking about times around one minute. Acceptance tests usually take a few of them, so the bottleneck won't be this but other workflows. My suggestion is to go ahead and implement it as a workflow that runs for every push.

> Alternatives

I really like the idea of generating a custom fixture for this purpose because we can generate them tailored to our test case, generating **large projects**. The projects that we have in the fixtures directories are tailored to test specific Tuist features, like the generation of a project with xcframeworks, or the definition of schemes.

> How we teach this

:+1: to performance guidelines. For example, if we believe complex glob patterns is a common source of slowdowns, against we can't do much but rather, ask the user to simplify them we can tell them in the guidelines.

> I’m personally not as familiar with GitHub Actions so will be learning as I go, of course contributions / assistance with this would be much appreciated and welcome.

If you are already familiar with other CI services, it's pretty much the same but with other syntax.

> Are there any existing benchmarking tools we could leverage?

Benchmarking is not a territory that I've explored a lot in the past. Maybe @RomainBoulay / @andreacipriani have some some of it at SoundCloud?

**Comment by vytis** [comment]
I agree that performance testing be very useful to do and could uncover unexpected regressions (or places to improve). Having a fixture generator is a good idea, but as with all synthetic benchmarks it is possible that it would not reflect real use cases.

The existing fixtures seem to exercise most of the features of Tuist and might be a good candidate for benchmarks. It also has the advantage that adding a new feature (and a fixture) would automatically add it to benchmarking suite.

When doing benchmarking I tend to use [hyperfine](https://github.com/sharkdp/hyperfine) which is very easy to use and can be customized easily. It also supports JSON output.

I tried benchmarking `tuist generate` in each fixture folder and there seems to be quite a lot of variation between them. These are the numbers (for 10 runs):

| Command | Mean [s] | Min [s] | Max [s] |
|:---|---:|---:|---:|
| `ios_app_with_incompatible_xcode` | 0.249 | 0.240 | 0.259 |
| `ios_app_with_actions` | 0.309 | 0.295 | 0.325 |
| `ios_app_with_extensions` | 0.326 | 0.318 | 0.336   |
| `ios_app_with_xcframeworks` | 0.343 | 0.334 | 0.355 |
| `ios_app_with_watchapp2` | 0.349 | 0.336 | 0.377 |
| `ios_app_with_pods` | 0.385 | 0.377 | 0.393  |
| `ios_app_with_carthage_frameworks` | 0.740 | 0.729 | 0.752 |
| `ios_app_with_setup` | 0.754 | 0.741 | 0.767 |
| `ios_app_with_tests` | 0.756 | 0.733 | 0.797  |
| `app_with_framework_and_tests` | 1.007 | 0.714 | 1.269  |
| `ios_app_with_sdk` | 1.243 | 1.207 | 1.356 |
| `ios_app_with_static_library_and_package` | 1.333 | 1.310 | 1.364 |
| `ios_app_with_remote_swift_package` | 1.381 | 1.350 | 1.421 |
| `ios_app_with_local_swift_package` | 1.559 | 1.533 | 1.580  |
| `ios_app_with_static_libraries` | 1.766 | 1.687 | 1.878 |
| `ios_app_with_static_frameworks` | 2.633 | 2.613 | 2.711 |

**Comment by kwridan** [comment]
Thanks both for your feedback!

> Having a fixture generator is a good idea, but as with all synthetic benchmarks it is possible that it would not reflect real use cases.
> The existing fixtures seem to exercise most of the features of Tuist and might be a good candidate for benchmarks. It also has the advantage that adding a new feature (and a fixture) would automatically add it to benchmarking suite.

Agreed leveraging existing fixtures will be useful to gather wide coverage. There is value in having both the existing fixtures and also some large fixtures too to help with profiling and highlight hotspots that may go unnoticed in the existing fixtures which are relatively small in size compared to typical projects. 

> When doing benchmarking I tend to use hyperfine which is very easy to use and can be customized easily. It also supports JSON output.

Ah neat! will check it out. Do you know if it has a way to give deltas by comparing two reports? Had a quick scan of the README may have missed it.

**Comment by vytis** [comment]
> Ah neat! will check it out. Do you know if it has a way to give deltas by comparing two reports? Had a quick scan of the README may have missed it.

I don't think so, it doesn't seem to be built exactly for the usecase we have. However to be able to track performance between commits we will have to somehow save the values to a database. I was thinking of something simple on Google Cloud since we already have it setup. It should work in a similar logic to what Codecov does - save values per commit and compare it between branches/PRs and `master`.
